blocks/api/categories.js

```diff
@@ -21,10 +24,121 @@ const categories = [
 ];
 
 /**
- * Returns all the block categories
+ * @type {RegExp}
+ * @const
+ *
+ * Category names must be a combination of lower-case letters, numbers, and hypens
+ */
+const categoryNamePattern = /^[a-z0-9-]+$/;
+
+/**
+ * Returns all the block categories.
  *
- * @return {Array} Block categories
+ * @return {Array} categories
  */
 export function getCategories() {
 	return categories;
 }
+
+/**
+ * Register a new block Category.
+ *
+ * @param {Array} category e.g {slug: 'custom', title: __('Custom Blocks')}
+ *
+ * @return {Array} categories
+ */
+export function registerCategory( category ) {
+	if ( ! category ) {
+		console.error(
+			'The block Category must be defined'
+		);
+		return;
+	}
+	if ( ! category.slug ) {
+		console.error(
+			'The block Category slug must be defined'
+		);
+		return;
+	}
+	if ( ! categoryNamePattern.test( category.slug ) ) {
+		console.error(
+			'The block Category slug must not contain characters which are invalid for urls'
+		);
+		return;
+	}
+
+	if ( categories.some( x => x.slug === category.slug ) ) {
+		console.error(
+			'The block Category "' + category.slug + '" is already registered'
+		);
+		return;
+	}
+	if ( ! category.title ) {
+		console.error(
+			'The block Category title must be defined'
+		);
+		return;
+	}
+
+	categories.push( category );
+	return categories;
+}
+
+/**
+ *
+ * Get sorted categories by property.
+ *
+ * @param {String} sortProperty The key to sort by
+ *
+ * @returns {Array} categories
+ */
+
+export function getSortedCategories( sortProperty ) {
+	if ( ! sortProperty ) {
+		console.error(
+			'The sortProperty must be defined'
+		);
+		return;
+	}
+	if ( typeof sortProperty !== 'string' ) {
+		console.error(
+			'The sortProperty must be a string'
+		);
+		return;
+	}
+
+	const sortedCategories = sortBy( categories, sortProperty );
+	return sortedCategories;
+}
```

Comment:
is this function even used? I didn't see any usage other than in the tests, plus, it's not really doing much

is there a reason we would prefer to call this over directly calling `sortBy()`?

I would imagine that this function probably isn't needed and we could strip it out

Reply:
@dmsnell Yes. The function is even more useless considering that we are not able to mutate categories. That's the reason  why it was a reassignable variable in one of my previous commit. I've tried to keep it simple but I agree with you it's better to have an isolated PR for this. 
It needs more discussions.